### PR TITLE
Fix SQLite scroll cursors fail

### LINKED ISSUE #35: PDO + SQLite + CURSOR_SCROLL failed
Hello :-),

Since our recent features on having iterators over result set, we force `PDO::CURSOR_SCROLL` on an `ATTR_CURSOR`. With SQLite, it failed with a null error. It seems to be an error from PDO but we have to fix it anyway.

Any idea how?

### PR DESCRIPTION
**Pull request by Metalaka**
Fix #35 

Hi,

I don't really like the `getAttribute` call but I don't see any other way to fix that properly.

=====
NB: We need `Hoa\Protocol` since we do a call to `resolve` in `Dal.php`.

### CODE REVIEW

File: Layer/Pdo/Pdo.php [hunk 1]
```diff
@@ -194,7 +194,12 @@ public function lastInsertId($name = null)
     public function prepare($statement, array $options = [])
     {
         if (!isset($options[\PDO::ATTR_CURSOR])) {
-            $options[\PDO::ATTR_CURSOR] = \PDO::CURSOR_SCROLL;
+            try {
+                $this->getConnection()->getAttribute(\PDO::ATTR_CURSOR);
+                $options[\PDO::ATTR_CURSOR] = \PDO::CURSOR_SCROLL;
```
Comment (Hywan) [inline Layer/Pdo/Pdo.php]: Actually, the `ATTR_CURSOR` is supported, but not the `CURSOR_SCROLL` value.

File: composer.json [hunk 2]
```diff
@@ -26,6 +26,7 @@
         "hoa/event"      : "~1.0",
         "hoa/exception"  : "~1.0",
         "hoa/iterator"   : "~2.0",
+        "hoa/protocol"   : "~1.0",
```
Comment (Hywan) [inline composer.json]: :+1:

File: Layer/Pdo/Pdo.php (hunk 1 above)
Comment (Metalaka) [inline Layer/Pdo/Pdo.php]: Did you try with this branch ?

I had the same error as you and this patch fix it (with sqlite3 on win10 x64)

Comment (Hywan) [inline Layer/Pdo/Pdo.php]: Hmm, I will try. However, iterator support will no be longer be available if scroll cursor is not enabled. Should we throw an error then?

Comment (Hywan) [inline Layer/Pdo/Pdo.php]: Also, if this patch works, then this is a bug in PDO with SQLite driver and we should open a bug on bugs.php.net, do you agree?

Comment (Metalaka) [inline Layer/Pdo/Pdo.php]: > Also, if this patch works, then this is a bug in PDO with SQLite driver and we should open a bug on bugs.php.net, do you agree?

Yes.

> Hmm, I will try. However, iterator support will no be longer be available if scroll cursor is not enabled. Should we throw an error then?

Forward only iterator will be available. Scrollable cursor is used to customize the fetch orientation.

Throw an error if the user want to use a feature which is not supported by the driver is interesting but we have to re-think the way we prepare our statement.
Currently we enable the scroll cursor before the user has chosen if he want to use the feature so we can't throw an error according to his choice.

Comment (Hywan) [inline Layer/Pdo/Pdo.php]: Agree with your strategy. Are you willing to do a PR?

File: Layer/Pdo/Pdo.php [hunk 3]
```diff
@@ -194,7 +194,12 @@ public function lastInsertId($name = null)
...
+            } catch (\PDOException $e) {
+                // Cursors isn't supported by the driver, see issue #35.
```
Comment (Hywan) [inline Layer/Pdo/Pdo.php]:
```diff
- isn't
+ are not
```
Fixed by me.

**Comment by Hywan** [comment]
Thanks for the PR!

**Pull request merged**